drivers/sensor/memsic/mmc56x3/mmc56x3_decoder.c

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2024 Intel Corporation
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "mmc56x3.h"
+#include <math.h>
+
+static int mmc56x3_decoder_get_frame_count(const uint8_t *buffer,
+					     struct sensor_chan_spec chan_spec,
+					     uint16_t *frame_count)
+{
+	const struct mmc56x3_encoded_data *edata = (const struct mmc56x3_encoded_data *) buffer;
+	int32_t ret = -ENOTSUP;
+
+	if (chan_spec.chan_idx != 0) {
+		return ret;
+	}
+
+	/* This sensor lacks a FIFO; there will always only be one frame at a time. */
+	switch (chan_spec.chan_type) {
+	case SENSOR_CHAN_AMBIENT_TEMP:
+		*frame_count = edata->has_temp ? 1 : 0;
+		break;
+	case SENSOR_CHAN_MAGN_X:
+		*frame_count = edata->has_magn_x ? 1 : 0;
+		break;
+	case SENSOR_CHAN_MAGN_Y:
+		*frame_count = edata->has_magn_y ? 1 : 0;
+		break;
+	case SENSOR_CHAN_MAGN_Z:
+		*frame_count = edata->has_magn_z ? 1 : 0;
+		break;
+	case SENSOR_CHAN_MAGN_XYZ:
+		*frame_count = ((edata->has_magn_x &&
+			edata->has_magn_y && edata->has_magn_z) ? 1 : 0);
+		break;
+	default:
+		return ret;
+	}
+
+	return 0;
+}
+
+static int mmc56x3_decoder_get_size_info(struct sensor_chan_spec chan_spec, size_t *base_size,
+					   size_t *frame_size)
+{
+	switch (chan_spec.chan_type) {
+	case SENSOR_CHAN_AMBIENT_TEMP:
+		*base_size = sizeof(struct sensor_q31_sample_data);
+		*frame_size = sizeof(struct sensor_q31_sample_data);
+	case SENSOR_CHAN_MAGN_X:
+	case SENSOR_CHAN_MAGN_Y:
+	case SENSOR_CHAN_MAGN_Z:
+	case SENSOR_CHAN_MAGN_XYZ:
+		*base_size = sizeof(struct sensor_three_axis_data);
+		*frame_size = sizeof(struct sensor_three_axis_sample_data);
+		return 0;
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static int mmc56x3_decoder_decode(const uint8_t *buffer, struct sensor_chan_spec chan_spec,
+				    uint32_t *fit, uint16_t max_count, void *data_out)
+{
+	const struct mmc56x3_encoded_data *edata = (const struct mmc56x3_encoded_data *)buffer;
+	const struct mmc56x3_data *data = &edata->data;
+
+	if (*fit != 0) {
+		return 0;
+	}
+
+	switch (chan_spec.chan_type) {
+	case SENSOR_CHAN_AMBIENT_TEMP:
+		struct sensor_q31_data *out = data_out;
+
+		out->header.base_timestamp_ns = edata->header.timestamp;
+		out->header.reading_count = 1;
+		if (edata->has_temp) {
+			double temp = MMC56X3_TEMP_BASE + (double) data->temp * MMC56X3_TEMP_RES;
```

Comment:
There should be a simple integer multiplier like the magnetometer channels to convert to q31

Reply:
Fixed, though I needed to add an offset